File: cpp/547.number-of-provinces.cpp
class UF {
private:
    int size;
    vector<int> parent;

public:
    UF(int _size) : size(_size), parent(vector<int>(size, 0))  {
        std::iota(parent.begin(), parent.end(), 0);
    }

    int find(int x) {
        while (x != parent[x]) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }

    bool join(int x, int y) {
        auto par_x = find(x), par_y = find(y);

        if (par_x == par_y) {
            return false;
        }
        parent[par_y] = par_x;
        return true;
    }

    int count() {
        int res = 0;
        vector<bool> vis(size, false);
        for (int v = 0; v < size; v++) {
            auto par_v = find(v);
            if (vis[par_v]) {
                continue;
            }
            vis[par_v] = true;
            res++;
        }
        return res;
    }

};

class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        UF uf(n);
        for (int u = 0; u < n; u++) {
            for (int v = u + 1; v < n; v++) {
                if (isConnected[u][v]) {
                    uf.join(u, v);
                }
            }
        }
        return uf.count();
    }
};

File: golang/547.number-of-provinces.go
func bfs(start int, vis []bool, gr [][]int) {
    q := make([]int, 0)
    q = append(q, start)
    vis[start] = true

    for len(q) > 0 {
        for m := len(q); m > 0; m-- {
            u := q[0]
            q = q[1:]
            for _, v := range gr[u] {
                if vis[v] {
                    continue
                }
                vis[v] = true
                q = append(q, v)
            }
        }
    }
}

func findCircleNum(isConnected [][]int) int {
    n := len(isConnected)
    gr := make([][]int, n)
    for i := 0; i < n; i++ { gr[i] = make([]int, 0) }

    for u := 0; u < n; u++ {
        for v := u + 1; v < n; v++ {
            if isConnected[u][v] == 1 {
                gr[u] = append(gr[u], v)
                gr[v] = append(gr[v], u)
            }
        }
    }

    vis := make([]bool, n)
    count := 0
    for u := 0; u < n; u++ {
        if vis[u] {
            continue
        }
        bfs(u, vis, gr)
        count++
    }
    return count;
}